Added CertificateSubject feature to xDSCWebService DSC Resource

rcarpenter79 opened the pull request:
Changed CertificateThumbprint from mandatory to use a default value AllowUnencryptedTraffic.
CertificateSubject can take a regex expression.
If multiple certificates are found then return the one with the latest not after date.
If no certificate subject is found set thumbprint to 'AllowUnencryptedTraffic'.

Fixes #57 
Fixes #205 
Fixes #418 

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379)
<!-- Reviewable:end -->

mgreenegit commented:
Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion.

---

*[DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1, line 849 at r2](https://reviewable.io:443/reviews/powershell/xpsdesiredstateconfiguration/379#-L256cc9Qtsn0UJQaBKH:-L256ccAlTMiM-3RtKa-:bb26vdb) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/9a46824b98fa6e008052e75b2276311ae085337e/DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1#L849)):*
> ```PowerShell
>         # Should execution stop if no certificate is found?
>         Write-Warning -Message "Certificate not found with subject containing $Subject and using template $TemplateName."
>         return 'AllowUnencryptedTraffic'
> ```

I think it would be better for the configuration to fail if no cert is found than to set the machine not in the desired state with a warning.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xpsdesiredstateconfiguration/379)*
<!-- Sent from Reviewable.io -->

mgreenegit commented:
Reviewed 1 of 2 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xpsdesiredstateconfiguration/379)*
<!-- Sent from Reviewable.io -->

mgreenegit commented:
I was just looking at submitting the same type of changes.  It actually used to work this way (back in 2014).  Thank you for submitting!!!  I submitted feedback.

rcarpenter79 commented:
Review status: 1 of 3 files reviewed at latest revision, 1 unresolved discussion.

---

*[DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1, line 849 at r2](https://reviewable.io:443/reviews/powershell/xpsdesiredstateconfiguration/379#-L256cc9Qtsn0UJQaBKH:-L25fwqt_yQ8uCpv1tI8:b2rgltl) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/9a46824b98fa6e008052e75b2276311ae085337e/DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1#L849)):*
<details><summary><i>Previously, mgreenegit (Michael Greene) wrote…</i></summary><blockquote>

I think it would be better for the configuration to fail if no cert is found than to set the machine not in the desired state with a warning.
</blockquote></details>

I was in two minds about that behavior. I have made the suggested change.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xpsdesiredstateconfiguration/379)*
<!-- Sent from Reviewable.io -->

kwirkykat commented:
@BerheAbrha or @Indhukrishna Can you take a look at this PR?

rcarpenter79 commented on DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1:
```diff
@@ -190,7 +232,11 @@ function Set-TargetResource
 
-    $os = [System.Environment]::OSVersion.Version
+    #$os = [System.Environment]::OSVersion.Version
+    $os = Get-CimInstance -ClassName Win32_OperatingSystem | Select-Object -Property `
```
I made this change to mock the OS version for testing.
Is there another way to do this without changing the code?

rcarpenter79 commented on DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1:
```diff
@@ -697,9 +801,9 @@ function Update-LocationTagInApplicationHostConfigForAuthentication
-    [System.Reflection.Assembly]::LoadWithPartialName("Microsoft.Web.Administration") | Out-Null
+    Add-Type -AssemblyName "Microsoft.Web.Administration, Version=7.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL"
 
-    $webAdminSrvMgr = [Microsoft.Web.Administration.ServerManager]::OpenRemote("127.0.0.1")
```
This was replaced with New-Object -TypeName Microsoft.Web.Administration.ServerManager so it can be mocked.

rcarpenter79 commented on DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1:
```diff
@@ -319,10 +365,7 @@ function Set-TargetResource
     }
     else
     {
-        if($AcceptSelfSignedCertificates -and ($AcceptSelfSignedCertificates -eq $false))
```
The If block will never be true so the code would never run.
I think I made a mistake here as I intended to comment out the line so it would be brought up to discuss the purpose.
I will change the If block to trigger only when $AcceptSelfSignedCertificates is $false

rcarpenter79 replied:
I had second thoughts on making this change. 
Wouldn't fixing the If statement result in a breaking change?
Would it be better to put it back how it was and add a comment for it to be checked if it's still needed?

BerheAbrha commented:
User need to explicitly opt-out to allow unencrypted traffic for security reasons.  I am not sure if making this optional is a good idea security wise. @TravisEz13  could you please advise on this.

BerheAbrha commented on DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1:
```diff
@@ -709,4 +813,40 @@ function Update-LocationTagInApplicationHostConfigForAuthentication
+            $_.Oid.FriendlyName -eq 'Certificate Template Name'
+        }.Format($false) -eq $TemplateName
+    } | Sort-Object -Property 'NotAfter' -Descending | Select-Object -First 1
+
```
If more than one cert is found which satisfy the criteria. I think It should error out instead of choosing one based on the expiry date. The reason being, the same cert have to be trusted by all clients connecting to this server and we can't tell which cert is trusted by all clients connecting to this PullServer.

rcarpenter79 commented:
Review status: 1 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1, line 840 at r3](https://reviewable.io:443/reviews/powershell/xpsdesiredstateconfiguration/379#-L-ktBVK8zIxNsECd4HD-r3-840:-L62sTStUaK6Y_5qAhBd:b-kbae3g) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/dda77097ab3704cc6bd398ff62af5783184a3d5b/DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1#L840)):*
<details><summary><i>Previously, TravisEz13 (Travis Plunk) wrote…</i></summary><blockquote>

Yes, there is a requirement that the user explicitly say that they intended for the endpoint to `AllowUnencryptedTraffic`
</blockquote></details>

I have updated the function to only allow one certificate to be returned.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xpsdesiredstateconfiguration/379)*
<!-- Sent from Reviewable.io -->

johlju commented:
@rcarpenter79 are you able to continue the work on this PR? Happy to help you get this reviewed. Could you please rebase this PR against dev so that we get the latest changes?

rcarpenter79 commented:
I have re-based my branch and fixed some tests that were failing. One test is now failing but in an unrelated  module. What is the recommended action?
I have checked the original files and they aren't different to the files in my branch.

johlju commented:
@rcarpenter79 Not sure the rebase worked as expected because there a lot of changes this PR want to do that is already part of dev, like that changes in the [README.md](https://github.com/PowerShell/xPSDesiredStateConfiguration/pull/379/files#diff-ced8f74ed7d50480c2a70a339d4dc529).
Did you use `git rebase`? See how to rebase here https://github.com/PowerShell/DscResources/blob/master/GettingStartedWithGitHub.md#resolve-merge-conflicts.

Let's start first with the rebase again, and then see if the problem with the tests still persists.

rcarpenter79 commented:
I think I have fixed the rebase now as it's only showing the files I modified and added.

johlju commented:
Thanks! Rebase looking good. :slightly_smiling_face:  Could you look at the review comments, please go into Reviewable and write 'Done' when they are resolved. When these review comments are resolved I will review again.

---

Review status: 0 of 3 files reviewed at latest revision, 4 unresolved discussions.

---

*[DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1, line 236 at r3](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-L-ktBVK8zIxNsECd4HD-r3-236:-LCcRMw3327AMgZqTIVR:bt10n9d) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/dda77097ab3704cc6bd398ff62af5783184a3d5b/DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1#L236)):*
<details><summary><i>Previously, rcarpenter79 wrote…</i></summary><blockquote>

I made this change to mock the OS version for testing.
Is there another way to do this without changing the code?
</blockquote></details>

Could you instead move the original string into a helper function that wraps the original way of getting the OS version? That wrapper function could be changed to use another method in the future, if needed.

---

*[DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1, line 49 at r5](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-LCcVpqV-Mwtng4MyHkM:-LCcVpqV-Mwtng4MyHkN:bulq8tp) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/925f2208d07d9b7e51a30e47146d008127a26608/DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1#L49)):*
> ```
> 'CertificateThumbprint must contain a certificate thumbprint or "AllowUnencryptedTraffic".
> ```

Maybe: `'CertificateThumbprint must contain a certificate thumbprint, or "AllowUnencryptedTraffic" to opt-out from being secure.'`

Throughout. Or even better add this as a localized string so it can be reused.

---

*[DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1, line 138 at r5](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-LCcUwye0glXC_H9lTac:-LCcUwye0glXC_H9lTad:b-wv3u8f) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/925f2208d07d9b7e51a30e47146d008127a26608/DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1#L138)):*
> ```PowerShell
>     else
>     {
>         $Certificate = ([Array](Get-ChildItem -Path 'Cert:\LocalMachine\My\')).Where{$_.Thumbprint -eq $webBinding.CertificateHash}
> ```

This will only work if the thumbprint is set? What should happen if only subject is set?

---

*[DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1, line 575 at r5](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-LCcY-ls2cC8L_e6Lilk:-LCcY-ls2cC8L_e6Lill:b-bhvbb6) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/925f2208d07d9b7e51a30e47146d008127a26608/DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1#L575)):*
> ```PowerShell
>                 }
>             }
>             'CertificateSubject'
> ```

Shouldn't this also check if `$CertificateThumbPrint -eq 'AllowUnencryptedTraffic'`?

---

*[DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1, line 866 at r5](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-LCcTKzn8G2iE2Kzc262:-LCcTKzn8G2iE2Kzc263:b-oy9cx0) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/925f2208d07d9b7e51a30e47146d008127a26608/DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1#L866)):*
> ```PowerShell
>     )
> 
>     Add-Type -AssemblyName "Microsoft.Web.Administration, Version=7.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL"
> ```

Using version here, would that break this resource in the future if only a newer version is available?

---

*[DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1, line 913 at r5](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-LCcSBXU6iYC0dBdz1Cs:-LCcSBXU6iYC0dBdz1Ct:bi7vuns) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/925f2208d07d9b7e51a30e47146d008127a26608/DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1#L913)):*
> ```PowerShell
>     )
> 
>     [Array] $CertificatesFromTemplates = (Get-ChildItem -Path $Store).Where{$_.Extensions.Oid.Value -contains '1.3.6.1.4.1.311.20.2'}
> ```

Could we add a comment that describes what the OID number used here means?

---

*[DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.schema.mof, line 5 at r4](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-L-ktBVK8zIxNsECd4HE-r4-5:-LCcQAFE-r9RGjmeRVj6:b-3kbmyz) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/62493ddb3006dedc145fa07523ba4672ebd6d35f/DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.schema.mof#L5)):*
<details><summary><i>Previously, TravisEz13 (Travis Plunk) wrote…</i></summary><blockquote>

We still need some mechanism to require the user to explicitly opt-out of being secure
</blockquote></details>

Can we use a new Boolean parameter to opt-out - default is secure, and throws if no certificate is specified. To opt-out this new parameter must be set to `$true` in the configuration. This would remove the use of the string `'AllowUnencryptedTraffic'` on the parameter `CertificateThumbPrint`. This would be a breaking change for configurations that have opt-out.

---


*Comments from [Reviewable](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-:-LCcSy9t2WIQSWGotKVc:b-kyp0ri)*
<!-- Sent from Reviewable.io -->

rcarpenter79 commented:
I think I should have addressed everything.

---

Review status: 0 of 4 files reviewed at latest revision, 9 unresolved discussions.

---

*[DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1, line 849 at r2](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-L256cc9Qtsn0UJQaBKH:-LCccMB08WrB1hrCF5a6:b-896fix) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/9a46824b98fa6e008052e75b2276311ae085337e/DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1#L849)):*
<details><summary><i>Previously, rcarpenter79 wrote…</i></summary><blockquote>

I was in two minds about that behavior. I have made the suggested change.
</blockquote></details>

Done.

---

*[DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1, line 236 at r3](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-L-ktBVK8zIxNsECd4HD-r3-236:-LCcvbkv05O_K0oZ0IOf:b-896fix) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/dda77097ab3704cc6bd398ff62af5783184a3d5b/DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1#L236)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Could you instead move the original string into a helper function that wraps the original way of getting the OS version? That wrapper function could be changed to use another method in the future, if needed.
</blockquote></details>

Done.

---

*[DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1, line 840 at r3](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-L-ktBVK8zIxNsECd4HD-r3-840:-LCccUjJ9hrvfJu4z1nx:b-896fix) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/dda77097ab3704cc6bd398ff62af5783184a3d5b/DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1#L840)):*
<details><summary><i>Previously, rcarpenter79 wrote…</i></summary><blockquote>

I have updated the function to only allow one certificate to be returned.
</blockquote></details>

Done.

---

*[DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1, line 49 at r5](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-LCcVpqV-Mwtng4MyHkM:-LCdFqqt6L1FJ1akQSxw:b-896fix) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/925f2208d07d9b7e51a30e47146d008127a26608/DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1#L49)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> ```
> 'CertificateThumbprint must contain a certificate thumbprint or "AllowUnencryptedTraffic".
> ```

Maybe: `'CertificateThumbprint must contain a certificate thumbprint, or "AllowUnencryptedTraffic" to opt-out from being secure.'`

Throughout. Or even better add this as a localized string so it can be reused.
</blockquote></details>

Done.

---

*[DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1, line 138 at r5](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-LCcUwye0glXC_H9lTac:-LCctQh48SNg-cdKip_Y:b-89t5bz) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/925f2208d07d9b7e51a30e47146d008127a26608/DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1#L138)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

This will only work if the thumbprint is set? What should happen if only subject is set?
</blockquote></details>

I'm returning the current configuration in the Get-TargetResource method. 
If the website is configured it will find the certificate hash and then get the subject and template name from the certificate. If the website isn't configured then nothing will be returned.
I'm not sure ignore the current configuration if 'AllowUnencryptedTraffic' is set.
An alternative would be to only return 'AllowUnencryptedTraffic' if the current certificate hash is null  and 'AllowUnencryptedTraffic' has been specified.

---

*[DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1, line 575 at r5](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-LCcY-ls2cC8L_e6Lilk:-LCcwA_75AvEYlULBEVL:b-hdrq6e) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/925f2208d07d9b7e51a30e47146d008127a26608/DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1#L575)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Shouldn't this also check if `$CertificateThumbPrint -eq 'AllowUnencryptedTraffic'`?
</blockquote></details>

CertificateSubject and CertificateThumbprint are in different parameter sets. If both are specified it should error. I don't think I added tests for that so I'll add that them while I'm at it.

---

*[DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1, line 866 at r5](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-LCcTKzn8G2iE2Kzc262:-LCd3Yzg5vQfEvMn7QJF:ba1bqdi) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/925f2208d07d9b7e51a30e47146d008127a26608/DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1#L866)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Using version here, would that break this resource in the future if only a newer version is available?
</blockquote></details>

You're right. Hard coding a version isn't sensible. I will lookup the version instead in a function so I can Mock it for testing when the assembly isn't present.

---

*[DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1, line 913 at r5](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-LCcSBXU6iYC0dBdz1Cs:-LCd3v1G060iLGB4I0w5:b-h09xmq) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/925f2208d07d9b7e51a30e47146d008127a26608/DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1#L913)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Could we add a comment that describes what the OID number used here means?
</blockquote></details>

I've added the comment and refactored the code to allow for certificates that use 'Certificate Template Information' as well as Certificate Template Name.
I've removed the use of the Oid in favour of using the FriendlyName for consistency.

---

*[DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.schema.mof, line 5 at r4](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-L-ktBVK8zIxNsECd4HE-r4-5:-LCdTyPC8Lhvp9q0pWWe:bgbyl4u) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/62493ddb3006dedc145fa07523ba4672ebd6d35f/DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.schema.mof#L5)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Can we use a new Boolean parameter to opt-out - default is secure, and throws if no certificate is specified. To opt-out this new parameter must be set to `$true` in the configuration. This would remove the use of the string `'AllowUnencryptedTraffic'` on the parameter `CertificateThumbPrint`. This would be a breaking change for configurations that have opt-out.
</blockquote></details>

I put a check in place to make sure something was specified for CertificateThumbprint.
I agree that an opt out parameter would probably be better.
Would it be better to raise that as an enhancement for a future code update?

---


*Comments from [Reviewable](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-:-LCdUag28SghzmSrPkz6:bp10ysi)*
<!-- Sent from Reviewable.io -->

rcarpenter79 commented:
There is definitely something odd about the xWindowsProcess Integration Tests.
It has failed on log files again.
Can I ignore it or do I have to make a change to trigger another build?

johlju commented:
You can close and reopen the PR to kick off the tests again. But this time you can make a change, there are a file that is missing new line at the end; https://ci.appveyor.com/project/PowerShell/xpsdesiredstateconfiguration/build/6.0.1074.0?fullLog=true#L55

johlju commented:
Reviewed 1 of 3 files at r5, 2 of 3 files at r6, 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1, line 236 at r3](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-L-ktBVK8zIxNsECd4HD-r3-236:-LCnNsu42j-bPrOv5fxB:b6ayd2g) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/dda77097ab3704cc6bd398ff62af5783184a3d5b/DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1#L236)):*
<details><summary><i>Previously, rcarpenter79 wrote…</i></summary><blockquote>

Done.
</blockquote></details>

It's currently a (reported) bug with Reviewable that writing 'Done.' with fulls top make me not able to acknowledge the review comments. :/ Please write 'Done' here instead :slightly_smiling_face: 

---

*[DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1, line 138 at r5](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-LCcUwye0glXC_H9lTac:-LCnLYRd1ur3spJcYGC-:b-xau9i4) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/925f2208d07d9b7e51a30e47146d008127a26608/DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1#L138)):*
<details><summary><i>Previously, rcarpenter79 wrote…</i></summary><blockquote>

I'm returning the current configuration in the Get-TargetResource method. 
If the website is configured it will find the certificate hash and then get the subject and template name from the certificate. If the website isn't configured then nothing will be returned.
I'm not sure ignore the current configuration if 'AllowUnencryptedTraffic' is set.
An alternative would be to only return 'AllowUnencryptedTraffic' if the current certificate hash is null  and 'AllowUnencryptedTraffic' has been specified.
</blockquote></details>

It feels like I want it to return $null if no certificate was set. Feels strange to return 'AllowUnencryptedTraffic' as the thumbprint. Here it is another good example that this should have been a boolean instead to opt-out from being secure. `Unsecure = $true` would have been better in by opinion. 
Let's leave it as is for now.

---

*[DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1, line 6 at r7](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-LCnKIWE9UBprvyPEJkm:-LCnKIWE9UBprvyPEJkn:b78wpru) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/8b518b40b31680e26eaaeb030a4a9e77eeba2d5b/DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1#L6)):*
> ```
> $Culture
> ```

I think this should be `$script:culture` according to style guideline. Throughout.

---

*[DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1, line 132 at r7](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-LCnLNToEOMPFxiwlviN:-LCnLNTp5G9IgBH3a4CA:bdyippq) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/8b518b40b31680e26eaaeb030a4a9e77eeba2d5b/DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1#L132)):*
> ```
> $Output
> ```

$output (lowe-case 'o'). Throughout.

---

*[DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1, line 155 at r7](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-LCnLTGq2hurVO2hur9X:-LCnLTGq2hurVO2hur9Y:b-q5sx4v) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/8b518b40b31680e26eaaeb030a4a9e77eeba2d5b/DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1#L155)):*
> ```
> $Certificate
> ```

$certificate (lower-case 'c').

---

*[DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1, line 880 at r7](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-LCnObg_00ax32KOn_sQ:-LCnObg_00ax32KOn_sR:bdfsth0) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/8b518b40b31680e26eaaeb030a4a9e77eeba2d5b/DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1#L880)):*
> ```
> $GacAssemblyVersion
> ```

$gacAssemblyVersion (lower-case 'g'). Throughout.

---

*[DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1, line 932 at r7](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-LCnOunt84YRKN_IVmay:-LCnOunt84YRKN_IVmaz:b-rd2k09) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/8b518b40b31680e26eaaeb030a4a9e77eeba2d5b/DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1#L932)):*
> ```
> $FilteredCertificates
> ```

$filteredCertificates lower-case. We should use camelCase for variables except parameters that uses PascalCase.

Throughout. To many to comment on each :slightly_smiling_face: Though, you only need to do your code changes, but wouldn't mind the other changed as well.

---

*[DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.schema.mof, line 5 at r4](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-L-ktBVK8zIxNsECd4HE-r4-5:-LCnQDw_7CISIv3Fb-I4:b-mvj9yx) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/62493ddb3006dedc145fa07523ba4672ebd6d35f/DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.schema.mof#L5)):*
<details><summary><i>Previously, rcarpenter79 wrote…</i></summary><blockquote>

I put a check in place to make sure something was specified for CertificateThumbprint.
I agree that an opt out parameter would probably be better.
Would it be better to raise that as an enhancement for a future code update?
</blockquote></details>

Leave it as is for now and I will raise the question and see if we should change this now. 

---

*[DSCResources/MSFT_xDSCWebService/en-US/MSFT_xDSCWebService.psd1, line 5 at r7](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-LCnQk0578jvcHKv6gZo:-LCnQk0578jvcHKv6gZp:b-utb3yi) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/8b518b40b31680e26eaaeb030a4a9e77eeba2d5b/DSCResources/MSFT_xDSCWebService/en-US/MSFT_xDSCWebService.psd1#L5)):*
> ```PowerShell
>     ThrowCertificateThumbprint       = CertificateThumbprint must contain a certificate thumbprint, or "AllowUnencryptedTraffic" to opt-out from being secure.
>     ThrowUseSecurityBestPractice     = Error: Cannot use best practice security settings with unencrypted traffic. Please set UseSecurityBestPractices to $false or use a certificate to encrypt pull server traffic.
>     FindCertificateBySubjectMultiple = More than one certificate found with subject containing {0} and using template {1}.
> ```

Nitpick (non-blocking): Maybe use quotes around template name since it can be spaces in the names?  This and below if so.

---

*[Tests/Unit/MSFT_xDSCWebService.Tests.ps1, line 26 at r7](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-LCnRfFR9pmzu_IkItZ8:-LCnRfFR9pmzu_IkItZ9:b-fbnv85) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/8b518b40b31680e26eaaeb030a4a9e77eeba2d5b/Tests/Unit/MSFT_xDSCWebService.Tests.ps1#L26)):*
> ```
> $DSCResourceName
> ```

camelCase here as well. Throughout the tests.

---

*[Tests/Unit/MSFT_xDSCWebService.Tests.ps1, line 168 at r7](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-LCnRr9T1FrWJ-Tmuh_w:-LCnRr9T1FrWJ-Tmuh_x:bme6fdp) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/8b518b40b31680e26eaaeb030a4a9e77eeba2d5b/Tests/Unit/MSFT_xDSCWebService.Tests.ps1#L168)):*
> ```PowerShell
> 
>             Context -Name 'DSC Web Service is installed without certificate' -Fixture {
>                 $Result = Get-TargetResource @TestParameters
> ```

Does Pester catches this if it throws an error if this is run in the Context-block and not in the It-block? Should we add this call to each It-block instead?

---

*[Tests/Unit/MSFT_xDSCWebService.Tests.ps1, line 239 at r7](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-LCnSQs_4e9op8uVbflg:-LCnSQs_4e9op8uVbflh:btbjh9x) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/8b518b40b31680e26eaaeb030a4a9e77eeba2d5b/Tests/Unit/MSFT_xDSCWebService.Tests.ps1#L239)):*
> ```
> 'should
> ```

We usually write 'Should...' (upper 'S').  Throughout.

---

*[Tests/Unit/MSFT_xDSCWebService.Tests.ps1, line 264 at r7](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-LCnT5-89X-E2COCKd6J:-LCnT5-89X-E2COCKd6K:b-44elsb) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/8b518b40b31680e26eaaeb030a4a9e77eeba2d5b/Tests/Unit/MSFT_xDSCWebService.Tests.ps1#L264)):*
> ```
> -Exactly 1
> ```

`-Exactly -Times 1`. Throughout.

---

*[Tests/Unit/MSFT_xDSCWebService.Tests.ps1, line 1038 at r7](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-LCnTfdd3sFCJC6-64MQ:-LCnTfde3GfVHsgf6n8U:b-5u94u8) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/8b518b40b31680e26eaaeb030a4a9e77eeba2d5b/Tests/Unit/MSFT_xDSCWebService.Tests.ps1#L1038)):*
> ```PowerShell
>             }
>             It 'should return $false if Module is present in Web.config and not expected to be installed.' {
> # This test is failing with existing code. 
> ```

Is this debug/todo comments? This and the one below?

---


*Comments from [Reviewable](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379)*
<!-- Sent from Reviewable.io -->

johlju commented:
@TravisEz13 @kwirkykat @mgreenegit Could you acknowledge any review comments you have pending or comment if there are still issues. Thanks!

@TravisEz13 What do you think about we add an `Unsecure` boolean property to the schema instead of opting-out by adding the string 'AllowUnencryptedTraffic' as the thumbprint? That would be a breaking change only for those configurations that has opted-out from being secure.

rcarpenter79 commented:
Review status: 1 of 4 files reviewed at latest revision, 15 unresolved discussions.

---

*[DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1, line 236 at r3](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-L-ktBVK8zIxNsECd4HD-r3-236:-LColfoy7yqDjhNR8BKx:bb6un1j) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/dda77097ab3704cc6bd398ff62af5783184a3d5b/DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1#L236)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

It's currently a (reported) bug with Reviewable that writing 'Done.' with fulls top make me not able to acknowledge the review comments. :/ Please write 'Done' here instead :slightly_smiling_face: 
</blockquote></details>

Done

---

*[DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1, line 6 at r7](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-LCnKIWE9UBprvyPEJkm:-LColQmyDXfFd3HTWna8:bb6un1j) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/8b518b40b31680e26eaaeb030a4a9e77eeba2d5b/DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1#L6)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> ```
> $Culture
> ```

I think this should be `$script:culture` according to style guideline. Throughout.
</blockquote></details>

Done

---

*[DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1, line 132 at r7](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-LCnLNToEOMPFxiwlviN:-LColU4G6Z8NNpi0ChU8:bb6un1j) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/8b518b40b31680e26eaaeb030a4a9e77eeba2d5b/DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1#L132)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> ```
> $Output
> ```

$output (lowe-case 'o'). Throughout.
</blockquote></details>

Done

---

*[DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1, line 155 at r7](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-LCnLTGq2hurVO2hur9X:-LColZSB08NO7d0MRQ3k:bb6un1j) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/8b518b40b31680e26eaaeb030a4a9e77eeba2d5b/DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1#L155)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> ```
> $Certificate
> ```

$certificate (lower-case 'c').
</blockquote></details>

Done

---

*[DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1, line 880 at r7](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-LCnObg_00ax32KOn_sQ:-LColhgd0kt_kva6rY0C:bb6un1j) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/8b518b40b31680e26eaaeb030a4a9e77eeba2d5b/DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1#L880)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> ```
> $GacAssemblyVersion
> ```

$gacAssemblyVersion (lower-case 'g'). Throughout.
</blockquote></details>

Done

---

*[DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1, line 932 at r7](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-LCnOunt84YRKN_IVmay:-LColiyk3x_P5IkpcBad:bb6un1j) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/8b518b40b31680e26eaaeb030a4a9e77eeba2d5b/DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1#L932)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> ```
> $FilteredCertificates
> ```

$filteredCertificates lower-case. We should use camelCase for variables except parameters that uses PascalCase.

Throughout. To many to comment on each :slightly_smiling_face: Though, you only need to do your code changes, but wouldn't mind the other changed as well.
</blockquote></details>

Done

---

*[DSCResources/MSFT_xDSCWebService/en-US/MSFT_xDSCWebService.psd1, line 5 at r7](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-LCnQk0578jvcHKv6gZo:-LColoyTCdN7Gg25tzbG:bb6un1j) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/8b518b40b31680e26eaaeb030a4a9e77eeba2d5b/DSCResources/MSFT_xDSCWebService/en-US/MSFT_xDSCWebService.psd1#L5)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Nitpick (non-blocking): Maybe use quotes around template name since it can be spaces in the names?  This and below if so.
</blockquote></details>

Done

---

*[Tests/Unit/MSFT_xDSCWebService.Tests.ps1, line 26 at r7](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-LCnRfFR9pmzu_IkItZ8:-LColr9y--jeOS6gRwQI:bb6un1j) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/8b518b40b31680e26eaaeb030a4a9e77eeba2d5b/Tests/Unit/MSFT_xDSCWebService.Tests.ps1#L26)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> ```
> $DSCResourceName
> ```

camelCase here as well. Throughout the tests.
</blockquote></details>

Done

---

*[Tests/Unit/MSFT_xDSCWebService.Tests.ps1, line 168 at r7](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-LCnRr9T1FrWJ-Tmuh_w:-LCosRdy5zBToxMrobNN:b-djljk7) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/8b518b40b31680e26eaaeb030a4a9e77eeba2d5b/Tests/Unit/MSFT_xDSCWebService.Tests.ps1#L168)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Does Pester catches this if it throws an error if this is run in the Context-block and not in the It-block? Should we add this call to each It-block instead?
</blockquote></details>

It throws an error in the Describe block and then doesn't run the remaining tests. As the remaining tests all check the result this isn't  exactly undesirable but it doesn't look good.
I'll move the assignment to an It-block and assign the output to a variable in the Describe context.

---

*[Tests/Unit/MSFT_xDSCWebService.Tests.ps1, line 239 at r7](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-LCnSQs_4e9op8uVbflg:-LColvqK0qswjnLMzJEB:bb6un1j) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/8b518b40b31680e26eaaeb030a4a9e77eeba2d5b/Tests/Unit/MSFT_xDSCWebService.Tests.ps1#L239)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> ```
> 'should
> ```

We usually write 'Should...' (upper 'S').  Throughout.
</blockquote></details>

Done

---

*[Tests/Unit/MSFT_xDSCWebService.Tests.ps1, line 264 at r7](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-LCnT5-89X-E2COCKd6J:-LColzNo9kDcPU3-h9Q7:bb6un1j) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/8b518b40b31680e26eaaeb030a4a9e77eeba2d5b/Tests/Unit/MSFT_xDSCWebService.Tests.ps1#L264)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> ```
> -Exactly 1
> ```

`-Exactly -Times 1`. Throughout.
</blockquote></details>

Done

---

*[Tests/Unit/MSFT_xDSCWebService.Tests.ps1, line 1038 at r7](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-LCnTfdd3sFCJC6-64MQ:-LCom058506i8866SmQl:brz0dki) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/8b518b40b31680e26eaaeb030a4a9e77eeba2d5b/Tests/Unit/MSFT_xDSCWebService.Tests.ps1#L1038)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Is this debug/todo comments? This and the one below?
</blockquote></details>

The test fails with existing code. The module IISSelfSignedCertModule(32bit) is present in the web.config and the test should be false.
I'm expecting the function to return $false but instead it returns $true.
Should I simply fix this function and make the test active again?

---


*Comments from [Reviewable](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379)*
<!-- Sent from Reviewable.io -->

johlju commented:
Awesome work fixing some much style issue. Thank you! :smiley: 

---

Reviewed 3 of 3 files at r8.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[a discussion](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-LD13A962sCFq9Qu1iAx:-LD13A962sCFq9Qu1iAy:b-7dcwte) (no related file):*
Could you please add an descriptive entry, for each change/issue, to the Unreleased section of the change log in the file README.md? If an entry resolves an issue please reference the issue.

---

*[DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1, line 246 at r8](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-LD0yj7G6Sn7LnriR1d0:-LD0yj7G6Sn7LnriR1d1:b8zlyd0) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/f14f2270d84214a358014dfdd50a336abeeba014/DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1#L246)):*
> ```
> [Bool]
> ```

Suggest we set all `[Bool]` to `[Boolean]`. That is mostly used throughout the DSC Resource Kit. Unless there is a particular reason to use `[Bool]`? 

---

*[DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1, line 309 at r8](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-LD0yShfCvMQIHgnYuNc:-LD0yShfCvMQIHgnYuNd:b-crlkua) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/f14f2270d84214a358014dfdd50a336abeeba014/DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1#L309)):*
> ```PowerShell
>     $esedatabase = "$DatabasePath\Devices.edb"
> 
>     $script:culture = Get-Culture
> ```

Should this be `$script:culture`? Or is is it meant to be `$culture`? This and the one below.

---

*[DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1, line 787 at r8](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-LD1-1lTEmgCCKfw3aM3:-LD1-1lTEmgCCKfw3aM4:b-zfrbdd) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/f14f2270d84214a358014dfdd50a336abeeba014/DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1#L787)):*
> ```PowerShell
>     param
>     (
>         [Parameter(Mandatory = $true)]
> ```

Just want to verify. By setting these parameters to mandatory, since there are no integration tests, doe the unit test verify that these are always called with both properties? Same for all other helper functions that have the mandatory property changed for the parameters.

---

*[DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.schema.mof, line 5 at r4](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-L-ktBVK8zIxNsECd4HE-r4-5:-LD10dcqAG_CKguGSB1U:bj05dzi) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/62493ddb3006dedc145fa07523ba4672ebd6d35f/DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.schema.mof#L5)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Leave it as is for now and I will raise the question and see if we should change this now. 
</blockquote></details>

Would you mind creating an issue for this enhancement?

---

*[Tests/Unit/MSFT_xDSCWebService.Tests.ps1, line 1038 at r7](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-LCnTfdd3sFCJC6-64MQ:-LD11qlf-JUZ9Gbgu6q6:b-ikgln8) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/8b518b40b31680e26eaaeb030a4a9e77eeba2d5b/Tests/Unit/MSFT_xDSCWebService.Tests.ps1#L1038)):*
<details><summary><i>Previously, rcarpenter79 wrote…</i></summary><blockquote>

The test fails with existing code. The module IISSelfSignedCertModule(32bit) is present in the web.config and the test should be false.
I'm expecting the function to return $false but instead it returns $true.
Should I simply fix this function and make the test active again?
</blockquote></details>

Please submit an issue for this. After that, if you want to resolve the problem then please do. :slightly_smiling_face:  Do you think it is a breaking change?

---


*Comments from [Reviewable](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-:-LD12w0I6uZk1OONvpCf:b-v0xvxk)*
<!-- Sent from Reviewable.io -->

rcarpenter79 commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1, line 246 at r8](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-LD0yj7G6Sn7LnriR1d0:-LD2K1jf8B8uraRBWW4H:b-loxg8q) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/f14f2270d84214a358014dfdd50a336abeeba014/DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1#L246)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> ```
> [Bool]
> ```

Suggest we set all `[Bool]` to `[Boolean]`. That is mostly used throughout the DSC Resource Kit. Unless there is a particular reason to use `[Bool]`? 
</blockquote></details>

There's no specific reason to use Bool. I always use the type accelerator, where available, over the .Net class.
I'll change it back to be consistent with the rest of the resource kit.

---

*[DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1, line 309 at r8](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-LD0yShfCvMQIHgnYuNc:-LD2Kzd37iCKsVldBUim:bb6un1j) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/f14f2270d84214a358014dfdd50a336abeeba014/DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1#L309)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Should this be `$script:culture`? Or is is it meant to be `$culture`? This and the one below.
</blockquote></details>

Done

---

*[DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1, line 787 at r8](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-LD1-1lTEmgCCKfw3aM3:-LD2MZ9xEcBOhkpE_9A_:b-4grbjn) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/f14f2270d84214a358014dfdd50a336abeeba014/DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1#L787)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Just want to verify. By setting these parameters to mandatory, since there are no integration tests, doe the unit test verify that these are always called with both properties? Same for all other helper functions that have the mandatory property changed for the parameters.
</blockquote></details>

I checked each occurrence of each function to confirm every parameter is specified when the function is called.
Your comment made me think what would happen if the function was passed a null value or empty string as this would now cause an error that wouldn't have happened previously.
I checked each function to see where the data came from. Half the functions were passing mandatory parameters from the main Get, Set and Test resources.
The remaining 4 all pass a path derived from Get-Website. If this fails then the other functions will not be called.
I don't expect these changes to cause any difference in behaviour and each function requires each parameter in order to return a sensible value.
If you prefer to leave this alone I'll remove the Mandatory value for the original functions.

---


*Comments from [Reviewable](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379)*
<!-- Sent from Reviewable.io -->

johlju commented:
@rcarpenter79 I think you forgot to push the changes? Can't see any changes between my comment and your comments. Or am I missing something here? 🤔

rcarpenter79 commented:
I ran out of time. I have pushed the latest changes now.

---

Review status: 1 of 5 files reviewed at latest revision, 9 unresolved discussions.

---

*[a discussion](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-LD13A962sCFq9Qu1iAx:-LD6NAH8De390DLXTmoJ:bb6un1j) (no related file):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Could you please add an descriptive entry, for each change/issue, to the Unreleased section of the change log in the file README.md? If an entry resolves an issue please reference the issue.
</blockquote></details>

Done

---

*[DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1, line 246 at r8](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-LD0yj7G6Sn7LnriR1d0:-LD6Mc2y4HmBswxCoWFc:bb6un1j) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/f14f2270d84214a358014dfdd50a336abeeba014/DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1#L246)):*
<details><summary><i>Previously, rcarpenter79 wrote…</i></summary><blockquote>

There's no specific reason to use Bool. I always use the type accelerator, where available, over the .Net class.
I'll change it back to be consistent with the rest of the resource kit.
</blockquote></details>

Done

---

*[DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.schema.mof, line 5 at r4](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-L-ktBVK8zIxNsECd4HE-r4-5:-LD6H_vY2D9QM88kPidV:b4ncgdm) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/62493ddb3006dedc145fa07523ba4672ebd6d35f/DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.schema.mof#L5)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Would you mind creating an issue for this enhancement?
</blockquote></details>

Raised as #419.

---

*[Tests/Unit/MSFT_xDSCWebService.Tests.ps1, line 1038 at r7](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-LCnTfdd3sFCJC6-64MQ:-LD65VX3A5tnqkIIYHkW:b-dwncyd) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/8b518b40b31680e26eaaeb030a4a9e77eeba2d5b/Tests/Unit/MSFT_xDSCWebService.Tests.ps1#L1038)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Please submit an issue for this. After that, if you want to resolve the problem then please do. :slightly_smiling_face:  Do you think it is a breaking change?
</blockquote></details>

Raised as #418.
I don't think this is a breaking change because I don't think existing code could work with it the way it currently is.
I have made a change and it's now passing my test.

---


*Comments from [Reviewable](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-:-LD6NE276xQHXx-xtVN7:b-updm00)*
<!-- Sent from Reviewable.io -->

rcarpenter79 commented:
Done

---

Review status: 1 of 5 files reviewed at latest revision, 9 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-:-LD6P-tO670ruAx_mJtm:bb6un1j)*
<!-- Sent from Reviewable.io -->

rcarpenter79 commented:
Review status: 4 of 5 files reviewed at latest revision, 5 unresolved discussions.

---

*[DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1, line 309 at r8](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-LD0yShfCvMQIHgnYuNc:-LD6biCK4OxPaw7CbkYD:b87o8fl) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/f14f2270d84214a358014dfdd50a336abeeba014/DSCResources/MSFT_xDSCWebService/MSFT_xDSCWebService.psm1#L309)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Want to double-check. Not sure if done means this is correct as-is or if this was supposed to be changed? :slightly_smiling_face: 
</blockquote></details>

I misunderstood. I thought you were referencing the style guidelines.
You're right. This variable assignment isn't correct compared with the original assignment at the top of the script.
I fixed it, which is handy as the flaky tests failed again.

---


*Comments from [Reviewable](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379)*
<!-- Sent from Reviewable.io -->

johlju commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Awaiting confirmation from other reviewers for a while before merging.

---

Reviewed 1 of 1 files at r10.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-:-LD6rrEv9kj39mMkjgqm:bp5stph)*
<!-- Sent from Reviewable.io -->

johlju commented:
@rcarpenter79 Think I found the problem with the tests, I hope. Gonna merge https://github.com/PowerShell/xPSDesiredStateConfiguration/pull/421 once the tests passes.

johlju commented:
@rcarpenter79 Can you please rebase now after PR #421 was merged. Hope the tests will all pass now! 🙂

johlju commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

After rebase for the merging of a fix for randomly failing tests.

---

Reviewed 1 of 1 files at r11.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/379#-:-LD77uqn5G4xKMTwobnQ:b-wn4cop)*
<!-- Sent from Reviewable.io -->
